src/home_robot_sim/home_robot_sim/env/habitat_ovmm_env/habitat_ovmm_env.py

```diff
@@ -168,33 +182,69 @@ def _preprocess_goal(
+        elif goal_type == "recep":
             goal_name = self._rec_id_to_name_mapping[obs["goal_receptacle"][0]]
-            rec_goal_id = 3
-            obj_goal_id = None
-        return obj_goal_id, rec_goal_id, goal_name
+            end_rec_goal_id = 3
+        return obj_goal_id, start_rec_goal_id, end_rec_goal_id, goal_name
 
-    def _preprocess_action(self, action: home_robot.core.interfaces.Action) -> int:
+    def _preprocess_action(
+        self, action: home_robot.core.interfaces.Action, habitat_obs
+    ) -> int:
         # convert planner output to continuous Habitat actions
-        action_map = {
-            DiscreteNavigationAction.TURN_RIGHT: [0, 0, -1, -1],
-            DiscreteNavigationAction.MOVE_FORWARD: [0, 1, 0, -1],
-            DiscreteNavigationAction.TURN_LEFT: [0, 0, 1, -1],
-            DiscreteNavigationAction.STOP: [0, 0, 0, 1],
-        }
-        print(action)
-        cont_action = action_map[action]
-        actions = {
-            "action": ("arm_action", "base_velocity", "rearrange_stop"),
-            "action_args": {
-                "arm_action": np.array([cont_action[0]], dtype=np.float32),
-                "base_vel": np.array(cont_action[1:3], dtype=np.float32),
-                "rearrange_stop": np.array([cont_action[-1]], dtype=np.float32),
-            },
-        }
-        return actions
+        grip_action = -1
+        if (
+            habitat_obs["is_holding"][0] == 1
+            and action != DiscreteNavigationAction.DESNAP_OBJECT
+        ) or action == DiscreteNavigationAction.SNAP_OBJECT:
+            grip_action = 1
+
+        turn = 0
+        if action == DiscreteNavigationAction.TURN_RIGHT:
+            turn = -1
+        elif action == DiscreteNavigationAction.TURN_LEFT:
+            turn = 1
+
+        forward = float(action == DiscreteNavigationAction.MOVE_FORWARD)
+        face_arm = float(action == DiscreteNavigationAction.FACE_ARM) * 2 - 1
+        stop = float(action == DiscreteNavigationAction.STOP) * 2 - 1
+        reset_joints = float(action == DiscreteNavigationAction.RESET_JOINTS) * 2 - 1
+        extend_arm = float(action == DiscreteNavigationAction.EXTEND_ARM) * 2 - 1
+        arm_actions = [0] * 7
+        cont_action = arm_actions + [
+            grip_action,
+            forward,
+            turn,
+            extend_arm,
+            face_arm,
+            stop,
+            reset_joints,
+        ]
+        return np.array(cont_action, dtype=np.float32)
 
     def _process_info(self, info: Dict[str, Any]) -> Any:
         if info:
             self.visualizer.visualize(**info)
+
+    def apply_action(
+        self,
+        action: home_robot.core.interfaces.Action,
+        info: Optional[Dict[str, Any]] = None,
+    ):
+        if info is not None:
+            self._process_info(info)
+        habitat_action = self._preprocess_action(action, self._last_habitat_obs)
+        habitat_obs, _, dones, infos = self.habitat_env.step(habitat_action)
```

Comment:
when we do this - is there a way to make it support either discrete or continuous?

Reply:
Yes, currently `_preprocess_action` that is being called here converts the discrete action (from planner) to continuous action (for habitat). This happens [here](https://github.com/facebookresearch/home-robot/blob/2a3c92389ba3833145f8c52d2ac9c162b9c4c1b4/src/home_robot_sim/home_robot_sim/env/habitat_ovmm_env/habitat_ovmm_env.py#L178). 

So only the home-robot->habitat conversion logic will change if the planner now outputs a continuous waypoint. I will make the discrete/continuous configurable.